go-controller/pkg/ovn/controller/services/load_balancer.go

```diff
@@ -296,37 +308,51 @@ func buildPerNodeLBs(service *v1.Service, configs []lbConfig, nodes []nodeInfo)
 			for _, config := range configs {
 				vips := config.vips
 
+				// we start with the default case: all endpoints are equally probable targets
 				routerV4targetips := config.eps.V4IPs
 				routerV6targetips := config.eps.V6IPs
 				switchV4targetips := config.eps.V4IPs
 				switchV6targetips := config.eps.V6IPs
 
+				// construct switch targets for all the endpoints
+				switchV4Targets := ovnlb.JoinHostsPort(switchV4targetips, config.eps.Port)
+				switchV6Targets := ovnlb.JoinHostsPort(switchV6targetips, config.eps.Port)
+
+				if config.topologyAwareRouting {
+					if topologyAwareRoutingIsEnabledForLBConfig(config.eps, node) {
+						// filter out endpoints for this node's zone on both switches and routers.
+						routerV4targetips = config.eps.ZoneHints[node.nodeLabels[v1.LabelTopologyZone]].V4IPs.List()
+						routerV6targetips = config.eps.ZoneHints[node.nodeLabels[v1.LabelTopologyZone]].V6IPs.List()
+						switchV4Targets = ovnlb.JoinHostsPort(config.eps.ZoneHints[node.nodeLabels[v1.LabelTopologyZone]].V4IPs.List(), config.eps.Port)
+						switchV6Targets = ovnlb.JoinHostsPort(config.eps.ZoneHints[node.nodeLabels[v1.LabelTopologyZone]].V6IPs.List(), config.eps.Port)
+					}
+				}
 				if config.externalTrafficLocal {
-					// for ExternalTrafficPolicy=Local, remove non-local endpoints from the router/switch targets
-					// NOTE: on the switches, filtered eps are used only by masqueradeVIP
-					routerV4targetips = util.FilterIPsSlice(routerV4targetips, node.nodeSubnets(), true)
-					routerV6targetips = util.FilterIPsSlice(routerV6targetips, node.nodeSubnets(), true)
+					// NOTE1: on the routers, we need to redo the filtering done by topologyAwareHints since ETP=local takes precendence
+					routerV4targetips = util.FilterIPsSlice(config.eps.V4IPs, node.nodeSubnets(), true)
+					routerV6targetips = util.FilterIPsSlice(config.eps.V6IPs, node.nodeSubnets(), true)
+					// NOTE1: on the switches, filtered eps are used only by masqueradeVIP
 					switchV4targetips = util.FilterIPsSlice(switchV4targetips, node.nodeSubnets(), true)
 					switchV6targetips = util.FilterIPsSlice(switchV6targetips, node.nodeSubnets(), true)
+					// NOTE3: on the switches, we need to reset the filtering done by topologyAwareHints since ETP=local takes precendence
+					switchV4Targets = ovnlb.JoinHostsPort(config.eps.V4IPs, config.eps.Port)
+					switchV6Targets = ovnlb.JoinHostsPort(config.eps.V6IPs, config.eps.Port)
 				}
```

Comment:
This whole thing of how we set an initial value on variables, change it for topologyAwareRouting and then reset it back to it original value if ETP is  very confusing. I suggest
```
if config.externalTrafficLocal {
 ...
} else if config.topologyAwareRouting && topologyAwareRoutingIsEnabledForLBConfig(config.eps, node) {
 ...
}
```

Alternatively or additionally use different variable names for the different scenarios (consider variables constant so set their values only once on initialization and use new variables instead of overriding) for better readibility.